alchemist/alchemist-swingui/src/main/java/it/unibo/alchemist/boundary/gui/effects/DrawLayers.java

```diff
@@ -0,0 +1,292 @@
+    @ExportForGUI(nameToExport = "A")
+    private RangedInteger alpha = new RangedInteger(0, MAX_COLOUR_VALUE, MAX_COLOUR_VALUE / INITIAL_ALPHA_DIVIDER);
+    @ExportForGUI(nameToExport = "R")
+    private RangedInteger red = new RangedInteger(0, MAX_COLOUR_VALUE);
+    @ExportForGUI(nameToExport = "G")
+    private RangedInteger green = new RangedInteger(0, MAX_COLOUR_VALUE);
+    @ExportForGUI(nameToExport = "B")
+    private RangedInteger blue = new RangedInteger(0, MAX_COLOUR_VALUE, MAX_COLOUR_VALUE);
+    private Color colorCache = Color.BLUE;
+    @Nullable
+    @SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
+    private transient Molecule molecule;
+    @Nullable
+    @SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
+    private transient Object molStringCached;
+    @SuppressFBWarnings(value = "SE_TRANSIENT_FIELD_NOT_RESTORED", justification = "If null, it gets reinitialized anyway if needed")
+    private transient CollectionWithCurrentElement<String> prevIncarnation;
+    private transient Incarnation<?, ?> incarnation;
+    private transient Optional<Node> markerNode = Optional.empty();
+
+    /**
+     */
+    public DrawLayers() {
+        final Set<String> availableIncarnations = SupportedIncarnations.getAvailableIncarnations();
+        if (availableIncarnations.isEmpty()) {
+            throw new IllegalStateException(getClass().getSimpleName() + " can't work if no incarnation is available.");
+        }
+        curIncarnation = new ImmutableCollectionWithCurrentElement<>(availableIncarnations, availableIncarnations.stream().findAny().get());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @SuppressWarnings({"PMD.CompareObjectsWithEquals", "unchecked"})
+    @SuppressFBWarnings("ES_COMPARING_STRINGS_WITH_EQ")
+    @Override
+    public <T, P extends Position2D<P>> void apply(final Graphics2D g, final Node<T> n, final Environment<T, P> env, final IWormhole2D<P> wormhole) {
+        // if marker node is no longer in the environment or it is no longer displayed, we need to change it
+        if (markerNode.isPresent() && (!env.getNodes().contains(markerNode.get())
+                                        || !wormhole.isInsideView(wormhole.getViewPoint(env.getPosition((Node<T>) markerNode.get()))))) {
+            markerNode = Optional.empty();
+        }
+        if (markerNode.isEmpty()) {
+            markerNode = Optional.of(n);
+        }
+        if (markerNode.get() == n) { // at this point markerNode.isPresent() is always true, so we directly get it
+            if (layerFilter && (incarnation == null || curIncarnation != prevIncarnation || molString != molStringCached)) {
+                molStringCached = molString;
+                prevIncarnation = curIncarnation;
+                incarnation = SupportedIncarnations.get(curIncarnation.getCurrent())
+                        .orElseThrow(() -> new IllegalStateException(curIncarnation.getCurrent() + " is not a valid incarnation."));
```

Comment:
This is an horrible kludge and looks like it's copypasted from elsewhere.
Avoid copy and paste, we already have a zillion lines of code to maintain, so every time you feel like copying something, either use inheritance or composition (strategy).

Reply:
It is indeed copy pasted from the `drawShape` class, I thought it was a good way to manage incarnations. In this particular case, inheritance cannot be used for obvious reasons, wouldn't it be jarring to use composition for such a small thing? 

One alternative is to get the current incarnation from the environment and remove this code completely, as well as the gui controls allowing the user to specify a particular incarnation

Reply:
do the following:

1. incapsulate the behavior, so that both places use the same code
2. yes, change it so that the incarnation is deduced from the environment.

Reply:
The incarnation is now inferred from the environment in my code, I didn't incapsulate that behaviour in an object as it seemed overkill to me. We are talking about incapsulating a single line of code:
```java
env.getIncarnation().ifPresent(incarnation -> molecule = incarnation.createMolecule(molString));
```
If `@ExportForGUI` worked through composition, I could isolate not just that line of code, but all the related controls (such as `molString`, `molStringCached`, etc), which actually makes more sense